File: js/tests/StreamParser.test.js
import { describe, expect, test } from 'bun:test';
import {
  formatLinks,
  Parser,
  StreamParseError,
  StreamParser,
} from '../src/index.js';

const document = `first loves data
profile:
  name Ada
  note "line one
line two"
(nested:
  child value)
last sees first`;

const formatted = (links) => formatLinks(links);

describe('StreamParser', () => {
  test('matches the canonical parser when fed one symbol at a time', () => {
    const stream = new StreamParser();
    for (const character of document) stream.write(character);

    expect(formatted(stream.end())).toBe(
      formatted(new Parser().parse(document))
    );
  });

  test('matches the canonical parser when fed one line at a time', () => {
    const stream = new StreamParser();
    for (const line of document.match(/.*(?:\n|$)/g).filter(Boolean)) {
      stream.write(line);
    }

    expect(formatted(stream.end())).toBe(
      formatted(new Parser().parse(document))
    );
  });

  test('emits a record as soon as the next top-level record starts', () => {
    const stream = new StreamParser();
    const seen = [];
    stream.on('link', (link) => seen.push(link.toString()));

    expect(stream.write('first loves data\n')).toEqual([]);
    expect(stream.write('s').map(String)).toEqual(['(first loves data)']);
    expect(seen).toEqual(['(first loves data)']);
  });

  test('does not emit an indented record before its children arrive', () => {
    const stream = new StreamParser();
    const seen = [];
    stream.on('link', (link) => seen.push(link));

    for (const character of 'profile:\n  name Ada\n  role engineer\n') {
      stream.write(character);
    }
    expect(seen).toEqual([]);

    stream.write('n');
    expect(formatted(seen)).toBe('(profile: (name Ada) (role engineer))');
  });

  test('does not split multiline quotes or parenthesized contexts', () => {
    const stream = new StreamParser();
    const seen = [];
    stream.on('link', (link) => seen.push(link));

    stream.write('"first\nsecond"\n(nested:\n  value one)\n');
    expect(seen).toHaveLength(1);
    stream.write('n');
    expect(seen).toHaveLength(2);
  });

  test('accepts a final record without a newline', () => {
    const stream = new StreamParser();
    stream.write('only one record');
    expect(formatted(stream.end())).toBe('(only one record)');
  });

  test('reports positions relative to the whole stream', () => {
    const stream = new StreamParser();
    stream.on('error', () => {});
    stream.write('good link\nnext link\nb');

    let caught;
    try {
      stream.end('ad: value: nope');
    } catch (error) {
      caught = error;
    }
    expect(caught).toBeInstanceOf(StreamParseError);
    expect(caught.line).toBe(3);
    expect(caught.offset).toBeGreaterThan('good link\nnext link\n'.length);
  });

  test('can avoid retaining output for bounded-memory event processing', () => {
    const stream = new StreamParser({ collect: false });
    const seen = [];
    stream.on('link', (link) => seen.push(link.toString()));
    stream.write('one link\nt');
    expect(stream.drain()).toEqual([]);
    expect(stream.end('wo link').map(String)).toEqual(['(two link)']);
    expect(seen).toEqual(['(one link)', '(two link)']);
  });

  test('drains retained output and can be reset', () => {
    const stream = new StreamParser();
    stream.write('one link\nt');
    expect(stream.drain().map(String)).toEqual(['(one link)']);
    expect(stream.end('wo link').map(String)).toEqual(['(two link)']);

    stream.reset().write('fresh record');
    expect(stream.end().map(String)).toEqual(['(fresh record)']);
  });

  test('offers synchronous and asynchronous iterable adapters', async () => {
    expect(
      [...StreamParser.parse(['one link\n', 'two link'])].map(String)
    ).toEqual(['(one link)', '(two link)']);

    async function* chunks() {
      yield 'one link\n';
      yield 'two link';
    }
    const links = [];
    for await (const link of StreamParser.parseAsync(chunks())) {
      links.push(link.toString());
    }
    expect(links).toEqual(['(one link)', '(two link)']);
  });

  test('limits only the unresolved record, not total stream size', () => {
    const stream = new StreamParser({ maxBufferSize: 12, collect: false });
    stream.on('error', () => {});
    for (let index = 0; index < 100; index++) stream.write(`a${index}\n`);
    expect(stream.end().map(String)).toEqual(['(a99)']);

    const tooLarge = new StreamParser({ maxBufferSize: 4 });
    tooLarge.on('error', () => {});
    expect(() => tooLarge.write('12345')).toThrow(RangeError);
  });

  test('tracks the absolute stream position', () => {
    const stream = new StreamParser();
    stream.write('one\ntw');
    expect(stream.position()).toEqual({
      offset: 6,
      line: 2,
      column: 3,
      buffered: 2,
    });
  });
});

File: go/stream_parser_test.go
package lino

import (
	"iter"
	"slices"
	"testing"
)

const streamDocument = "first loves data\n# streamed comment\nprofile:\n  name Ada\n  note \"line one\nline two\"\n(nested:\n  child value)\nlast sees first"

func linkStrings(links []*Link) []string {
	result := make([]string, len(links))
	for index, link := range links {
		result[index] = link.String()
	}
	return result
}

func TestMatchesCanonicalParserOneSymbolAtATime(t *testing.T) {
	stream := NewStreamParser()
	for _, symbol := range streamDocument {
		if _, err := stream.Feed(string(symbol)); err != nil {
			t.Fatal(err)
		}
	}
	actual, err := stream.Finish()
	if err != nil {
		t.Fatal(err)
	}
	expected, err := Parse(streamDocument)
	if err != nil {
		t.Fatal(err)
	}
	if !slices.Equal(linkStrings(actual), linkStrings(expected)) {
		t.Fatalf("streamed links %v, want %v", linkStrings(actual), linkStrings(expected))
	}
}

func TestEmitsOnlyCompleteRecords(t *testing.T) {
	stream := NewStreamParser()
	var seen []string
	stream.OnLink = func(link *Link) { seen = append(seen, link.String()) }

	if emitted, err := stream.Feed("first loves data\n"); err != nil || len(emitted) != 0 {
		t.Fatalf("Feed() = %v, %v", emitted, err)
	}
	emitted, err := stream.Feed("s")
	if err != nil {
		t.Fatal(err)
	}
	if !slices.Equal(linkStrings(emitted), []string{"(first loves data)"}) || !slices.Equal(seen, linkStrings(emitted)) {
		t.Fatalf("emitted %v, callback %v", linkStrings(emitted), seen)
	}
}

func TestStreamParserPreservesIndentedAndMultilineRecords(t *testing.T) {
	stream := NewStreamParser()
	var seen []*Link
	stream.OnLink = func(link *Link) { seen = append(seen, link) }

	if _, err := stream.Feed("profile:\n  name Ada\n  note \"first\nsecond\"\n"); err != nil {
		t.Fatal(err)
	}
	if len(seen) != 0 {
		t.Fatalf("emitted incomplete record: %v", linkStrings(seen))
	}
	if _, err := stream.Feed("next"); err != nil {
		t.Fatal(err)
	}
	if len(seen) != 1 {
		t.Fatalf("emitted %d links, want 1", len(seen))
	}
}

func TestSupportsDrainResetAndBoundedMemory(t *testing.T) {
	stream := NewStreamParser()
	stream.Collect = false
	stream.MaxBufferSize = 12
	for index := range 100 {
		if _, err := stream.Write([]byte("a\n")); err != nil {
			t.Fatal(err)
		}
		if index == 50 {
			stream.Reset()
			stream.Collect = false
			stream.MaxBufferSize = 12
		}
	}
	if len(stream.Drain()) != 0 {
		t.Fatal("Collect=false retained links")
	}
	if _, err := stream.Finish(); err != nil {
		t.Fatal(err)
	}

	tooLarge := NewStreamParser()
	tooLarge.MaxBufferSize = 4
	if _, err := tooLarge.Feed("12345"); err == nil {
		t.Fatal("oversized unresolved record was accepted")
	}
}

func TestProvidesLazyAdapters(t *testing.T) {
	chunks := func(yield func(string) bool) {
		for _, chunk := range []string{"one link\n", "two link"} {
			if !yield(chunk) {
				return
			}
		}
	}
	var actual []string
	for link, err := range ParseChunks(iter.Seq[string](chunks), nil) {
		if err != nil {
			t.Fatal(err)
		}
		actual = append(actual, link.String())
	}
	if !slices.Equal(actual, []string{"(one link)", "(two link)"}) {
		t.Fatalf("ParseChunks() = %v", actual)
	}
}

func TestSupportsLineChunksFinalRecordAndPosition(t *testing.T) {
	stream := NewStreamParser()
	for _, line := range stringsAfter(streamDocument, '\n') {
		if _, err := stream.Feed(line); err != nil {
			t.Fatal(err)
		}
	}
	actual, err := stream.Finish()
	if err != nil {
		t.Fatal(err)
	}
	expected, err := Parse(streamDocument)
	if err != nil {
		t.Fatal(err)
	}
	if !slices.Equal(linkStrings(actual), linkStrings(expected)) {
		t.Fatalf("streamed links %v, want %v", linkStrings(actual), linkStrings(expected))
	}
	position := stream.Position()
	if position.Offset != len(streamDocument) || position.Buffered != 0 {
		t.Fatalf("Position() = %+v", position)
	}
}

func TestRejectsWritesAfterFinish(t *testing.T) {
	stream := NewStreamParser()
	if _, err := stream.Finish("one"); err != nil {
		t.Fatal(err)
	}
	if _, err := stream.Feed("two"); err == nil {
		t.Fatal("write after Finish was accepted")
	}
}

func stringsAfter(value string, delimiter byte) []string {
	var chunks []string
	start := 0
	for index := range len(value) {
		if value[index] == delimiter {
			chunks = append(chunks, value[start:index+1])
			start = index + 1
		}
	}
	if start < len(value) {
		chunks = append(chunks, value[start:])
	}
	return chunks
}
